### ISSUE TITLE
Infinite recursion when mutating `raise` calls within `#to_s`

### ISSUE THREAD

**Issue by jaynetics** [mbj/mutant#965]
mutant replaces the second argument to `raise` with `self`, leading to infinite recursion if that raise call is within `#to_s`, as raise tries to stringify that argument.

minimal repro:

```bash
echo 'class C; def to_s; raise TypeError, "foo" end end' > c.rb
echo 'RSpec.describe(C) { it("#to_s") { expect { C.new.to_s }.to raise_error(TypeError, "foo") } }' > spec/c_spec.rb
bundle exec mutant --require './c' --use rspec C
```

this one raises `ArgumentError: marshal data too short`, but I've also seen `EOFError`.

there might be other cases where inserting `self` within `#to_s` causes problems.

**Comment by mbj** [comment]
@jaynetics So the mutation itself is fine. Mutating to non terminating code is part of mutants job. Your test suite should detect these.

The problem is that mutant does not appear to handle the crash of the killfork correctly, and this is something to address.

thanks for the very short repro, this is something to work with!

**Comment by mbj** [comment]
@jaynetics Are you on the latest release? This one has various problems with good termination detection, I've got fixes in master that are under the new license.

If you wanted to test on master, feel free to tell me a repo name, so I can give you an opensource license key.

**Comment by tomdalling** [comment]
I can confirm that this is still happening for me. My `Gemfile.lock` versions are:

    mutant (0.9.8)
    mutant-license (0.1.0)
    mutant-rspec (0.9.5)

**Comment by mbj** [comment]
@tomdalling Can you post the output?

**Comment by tomdalling** [comment]
Here's a gist with the output from the minimal repro above, and also the output of the error that I ran into myself: https://gist.github.com/tomdalling/3f1fc7b2b20c6035023b063e882b3dcc

**Comment by mbj** [comment]
@tomdalling Hah thats an interesting one! Let me elaborate. The bug is in rspec ;) but to explain why I need to draw a bigger circle:

Mutant on inserting mutations, want to measure the effect of the mutation via the report provided by the test framework (in your case rspec). The report by the test framework, that may have indicated that a test failed, will than determine if the mutation is covered or not. 

You have hit a case where rspec does NOT produce a report and instead crashes. Hence mutant refuses to take this crash result as a covered mutation, for the reason it has to be assumed that "any exeuction that does not produce a report" is a bug in mutants code (or its dependencies, in this case rspec!) that executes the test.

This "integration error vs real test error" separation is very important, as the act of mutation testing itself, not the actual mutation (!!!), can make the coverage attempt fail. If mutant would count "all crashes of the test framework itself, or crashes of its own code" as covered mutations: Mutants reports would contain mutations as covered that actually are not if you where to manually insert them. As mutation testing puts the runtime system, the application code, and the test framework into "unusual" states this distinction found me many integration problems and dependency bugs. Including memory corruption in MRI ;)

So you found "another" of these cases. Cudos.

OT: Many mutation testing tools do NOT make this very important distinction between coverage by test framework report on failed tests and integration errors.

So how to move on, IMO its a bug in rspec, as rspec should never raise execptions coming from calling APIs that are controlled from the system under test.

your exceptions `#to_s` method gets tested, using it raises: Rspec should count this as a test failure.

If you agree with me follow this path:

* Reproduce the crash of rspec, outside of mutant, via applying the mutation manually and re-run rspec
* Report it to rspec and get it fixed. Rspec in this case should actually report a test failure and not exit without producing a test report. You can tell its their intention to produce a report from the location of the exception: `/Users/tom/.gem/ruby/2.7.1/gems/rspec-core-3.9.2/lib/rspec/core/formatters/exception_presenter.rb:180:in `to_s'` its already past executing your tests and just tries to format it.
* Update rspec ;)

Should the rspec people take some time, or refuse to fix it, these are your options:

* Keep the subject in your ignore list.
* Wait for a feature where I allow to whitelist certain subjects integration failure as expected and counting them as covered. I'll not switch to globally whitelist integration failures as covered as this would turn a blind eye to integration errors.

**Comment by tomdalling** [comment]
Thanks for the detailed explanation! I'll report it to RSpec and see what happens.

**Comment by mbj** [comment]
> Thanks for the detailed explanation! I'll report it to RSpec and see what happens.

Heh, I need to turn that explanation into a blog post: "why your mutation testing engine needs noop mutations and integration error separation". I saw code bases from 100% coverage go to 50% after adding these features to the engine.

**Comment by mbj** [comment]
@tomdalling BTW this is a "classic" case on mutant finding an imperfection in the general Ruby toolchain. And the core reason we have more than "covered" and "alive", but also "integration error" in mutant.

I see traffic over at the rspec ticket you made, thanks for getting it fixed upstream.

**Comment by tomdalling** [comment]
👍  I found another RSpec issue yesterday too, thanks to mutant.

**Comment by tomdalling** [comment]
This one is fixed for me after updating to `rspec-core` v3.9.3.

**Comment by mbj** [comment]
@tomdalling Okay, I'm closing this one. We have other instances of inifinte recursion crashing MRI itself, these will need to be handled differently than this one.

If you disagree this issue is resolved please comment and I'll re-open.
